### ISSUE TITLE
Log training losses at each step for multistep loss.

### ISSUE THREAD

**Issue by gideonite** [ai2cm/ace#212]
Currently we only track `batch_loss` but for multistep losses, it would make sense to track each individual loss as a sanity check.

**Comment by oliverwm1** [comment]
I agree this would be a useful thing to track (and I recall the original fourcastnet training code did track both of them).

**Comment by gideonite** [comment]
I think that this will require making changes through the aggregator code base. `record_batch` which normally takes a single `loss:float` as an argument will be changed to accept `losses: torch.Tensor`. When it comes to finally logging to wandb, I'll make a simple check to see if `n_forward_steps = 1`, if so, only log a single value, otherwise, log `loss_1`, `loss_2`, ...

Does this sound good to you? @oliverwm1 @mcgibbon

**Comment by mcgibbon** [comment]
Please always save the single value loss, in addition to any step-separated losses. I know these can be summed but it's a pain.

Are we talking about the batch loss or the mean loss? The aggregators do not handle the batch loss and shouldn't need to be updated for changes to the batch loss. For this I would suggest having run_on_batch return a dict of metrics that include the loss and these new step losses.

If you want this for mean losses, can we save the 2nd order LpLoss at each step with a new aggregator, instead of overhauling the api of every aggregator?

**Comment by gideonite** [comment]
My understanding is that we just want training loss, probably either batch loss or the mean loss would be fine. Mainly, this is a sanity check on the multistep training.

**Comment by mcgibbon** [comment]
"this is a sanity check on the multistep training" can you elaborate on what you mean?  If there is a concern as to whether the multi step loss is computed correctly, we should write a unit test of run_on_batch with a prespecified stepper (like the plus one stepper) and assert that its multistep loss is what we expect analytically. This test might already exist.

Is there a diagnostic purpose for this in wandb other than a sanity check? It seems helpful to me to be able to see if the losses are about what we expect (linearly increasing with step).

**Comment by gideonite** [comment]
I'm not exactly sure but I think that Chris wanted to make sure that the model is doing what we expect it to be doing.

Unit test makes sense as well, either way works for me. If you are referring to [plus one stepper](https://github.com/ai2cm/full-model/blob/093be949c1986ca3fe59dcceadae9f6f1afa8fd8/fme/fme/fcn_training/inference/test_inference.py#L25), it looks like this is only being used for testing stepper saving.

Let me know what you think would be the best way to proceed. @oliverwm1 @mcgibbon

**Comment by oliverwm1** [comment]
It may be worth a slightly more general view: if/when we add more terms to the loss (e.g. global mean L1 loss, or physical constraint loss) we will likely want to be able to track how these individual terms are changing through training. Not just as a sanity check, but to actually understand how the model is weighting the different parts of the loss.

In terms of implementation, I'll defer to @mcgibbon. But I agree it would be fine to just track for the batch loss, so simply returning it as output of `run_on_batch` and not touching the aggregators seems like a good approach.

**Comment by gideonite** [comment]
Sorry, I don't think I'm following this Oli. Sounds like at some point
we'll want to track different parts of the loss but at the moment it's fine
to just write a unit test to check that the autoregressive loss is being
computed correctly.

Is that right?

**Comment by mcgibbon** [comment]
" I think that Chris wanted to make sure that the model is doing what we
expect it to be doing."

I think for this purpose we should have a unit test. I'm not sure how we
would determine "what we expect it to be doing" for wandb losses.

I also agree with Oli about the general usefulness of having step losses as
a diagnostic, but it's separate from sanity checking / code testing.

Busy shuttling around hospitals today, but if we need to talk more I'll
follow up tomorrow.

**Comment by gideonite** [comment]
I agree with you. I'm going to go ahead and implement a unit test that does
that.

**Comment by oliverwm1** [comment]
I am +1 to having a unit test for multi-step loss if we don't have it already.

My understanding of what Matt/Chris wanted was that it was not simply a test of "is the code working" but also a plot that shows the values of each of the terms in the loss, and how they change during training.

**Comment by gideonite** [comment]
Works for me. I'm going to go ahead and do both. Sounds like I should create a separate code path for these losses aside from the loss. I'll update with a plan on how to achieve that later today.

**Comment by gideonite** [comment]
I think [this existing unit test](https://github.com/ai2cm/full-model/blob/main/fme/fme/core/test_stepper.py#L114) is sufficient actually.

**Comment by mcgibbon** [comment]
> I think [this existing unit test](https://github.com/ai2cm/full-model/blob/main/fme/fme/core/test_stepper.py#L114) is sufficient actually.

You may want to add assertions about the loss.
